File: backend/app/schemas/analytics.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class MetricSummary(BaseModel):
    current: float | None = None
    previous: float | None = None
    change_pct: float | None = None
    unit: str | None = None
    trend_direction: str | None = None  # "up" | "down" | "flat"
    min_value: float | None = None
    max_value: float | None = None
    avg_value: float | None = None


class TimeSeriesPoint(BaseModel):
    date: datetime
    value: float


class SiteComparisonItem(BaseModel):
    site_id: UUID
    site_name: str
    area_hectares: float | None = None
    metrics: dict[str, float | None] = {}


class SiteAnalytics(BaseModel):
    site_id: UUID
    site_name: str
    area_hectares: float | None = None
    metrics: dict[str, MetricSummary]
    time_series: dict[str, list[TimeSeriesPoint]]
    observation_count: int
    is_synthetic: bool = True
    data_source_note: str = "Synthetic Demo Dataset — Calibrated for Hackathon Demonstration"


class ProjectAnalytics(BaseModel):
    project_id: UUID
    project_name: str
    site_count: int
    total_area_hectares: float
    metrics: dict[str, MetricSummary]
    time_series: dict[str, list[TimeSeriesPoint]] = {}
    site_breakdown: list[SiteComparisonItem] = []
    is_synthetic: bool = True
    data_source_note: str = "Synthetic Demo Dataset — Calibrated for Hackathon Demonstration"


class DashboardKPIs(BaseModel):
    total_projects: int
    active_projects: int
    total_sites: int
    total_area_hectares: float
    avg_carbon_stock: float | None = None
    avg_biodiversity_index: float | None = None

File: frontend/src/types/analytics.ts
export interface MetricSummary {
  current: number | null
  previous: number | null
  change_pct: number | null
  unit: string | null
  trend_direction?: 'up' | 'down' | 'flat' | null
  min_value?: number | null
  max_value?: number | null
  avg_value?: number | null
}

export interface TimeSeriesPoint {
  date: string
  value: number
}

export interface SiteComparisonItem {
  site_id: string
  site_name: string
  area_hectares: number | null
  metrics: Record<string, number | null>
}

export interface SiteAnalytics {
  site_id: string
  site_name: string
  area_hectares: number | null
  metrics: Record<string, MetricSummary>
  time_series: Record<string, TimeSeriesPoint[]>
  observation_count: number
  is_synthetic: boolean
  data_source_note: string
}

export interface ProjectAnalytics {
  project_id: string
  project_name: string
  site_count: number
  total_area_hectares: number
  metrics: Record<string, MetricSummary>
  time_series: Record<string, TimeSeriesPoint[]>
  site_breakdown: SiteComparisonItem[]
  is_synthetic: boolean
  data_source_note: string
}

export interface DashboardKPIs {
  total_projects: number
  active_projects: number
  total_sites: number
  total_area_hectares: number
  avg_carbon_stock: number | null
  avg_biodiversity_index: number | null
}

export interface ObservationMetricItem {
  id: string
  metric_type: string
  value: number
  unit: string | null
}

export interface Observation {
  id: string
  site_id: string
  observed_at: string
  source: string
  source_reference: string | null
  created_at: string
  metrics: ObservationMetricItem[]
}

export interface Organization {
  id: string
  name: string
  slug: string
  created_at: string
}

export interface EnvironmentalContext {
  site_id: string
  latitude: number
  longitude: number
  weather: {
    temperature_celsius: number
    humidity_pct: number
    precipitation_mm: number
    wind_speed_kmh: number
    soil_moisture_pct: number
    condition: string
    source: string
    is_live: boolean
  }
  earth_observation: {
    satellite_ndvi: number
    canopy_cover_pct: number
    fire_risk_level: string
    fire_risk_index: number
    cloud_cover_pct: number
    satellite_source: string
    is_live: boolean
  }
}

export interface ObservationCreatePayload {
  observed_at: string
  source: string
  source_reference?: string
  metrics: Array<{
    metric_type: string
    value: number
    unit?: string
  }>
}

export interface ObservationFilterParams {
  start_date?: string
  end_date?: string
  metric_type?: string
  source?: string
  skip?: number
  limit?: number
}
